README.md
# Swift iOS ANE  XCode9 and iOS 11 Beta

Example Xcode project showing how to create Air Native Extensions for iOS using Swift 4.0.


----------

**This branch is experimental. It is used to build ANEs for iOS 11 Beta using XCode 9 beta.
Please see tech notes below. There is some configuration needed !!**

The example uses PDFKit with is new in iOS 11

----------

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=5UR2T52J633RC)

It is comprised of 3 parts.

1. A static library which exposes methods to AIR and a thin ObjectiveC API layer to the Swift code.
2. A dynamic Swift Framework which contains the translation of FlashRuntimeExtensions to Swift.
3. A dynamic Swift Framework which contains the main logic of the ANE.

> To allow FRE functions to be called from within Swift a protocol acting 
> as a bridge back to Objective C was used.

SwiftIOSANE_LIB/SwiftIOSANE_LIB.m is the entry point of the ANE. It acts as a thin layered API to your Swift controller.  
Add the number of methods here 

````objectivec
static FRENamedFunction extensionFunctions[] =
{
     MAP_FUNCTION(TRSOA, load)
    ,MAP_FUNCTION(TRSOA, goBack)
};
`````


SwiftIOSANE_FW/SwiftController.swift  
Add Swift method(s) to the functionsToSet Dictionary in getFunctions()

````swift
@objc public func getFunctions(prefix: String) -> Array<String> {
    functionsToSet["\(prefix)load"] = load
    functionsToSet["\(prefix)goBack"] = goBack    
}
`````

Add Swift method(s)

````swift
func load(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
    //your code here
    return nil
}

func goBack(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
    //your code here
    return nil
}
`````

----------

### How to use
######  The methods exposed by FlashRuntimeExtensions.swift are similar to the FreSharp API for Air Native Extensions. 

Example - Convert a FREObject into a String, and String into FREObject

````swift
let airString: String = FreObjectSwift(freObject: inFRE0).value as? String
trace("String passed from AIR:", airString)
let swiftString: String = "I am a string from Swift"
do {
    return try FreObjectSwift(string: swiftString).rawValue
} catch {}
`````


Example - Call a method on an FREObject

````swift
if let addition: FreObjectSwift = try person.callMethod(name: "add", args: 100, 31) {
    if let sum: Int = addition.value as? Int {
        trace("addition result:", sum)
    }
}
`````

Example - Reading items in array
````swift
let airArray: FreArraySwift = FreArraySwift.init(freObject: inFRE0)
do {
    if let itemZero: FreObjectSwift = try airArray.getObjectAt(index: 0) {
        if let itemZeroVal: Int = itemZero.value as? Int {
            trace("AIR Array elem at 0 type:", "value:", itemZeroVal)
            let newVal = try FreObjectSwift.init(int: 56)
            try airArray.setObjectAt(index: 0, object: newVal)
            return airArray.rawValue
         }
    }
} catch {}
`````

Example - Convert BitmapData to a UIImage and add to native view
````swift
let asBitmapData = FreBitmapDataSwift.init(freObject: inFRE0)
defer {
    asBitmapData.releaseData()
}
do {
    if let cgimg = try asBitmapData.getAsImage() {
        let img:UIImage = UIImage.init(cgImage: cgImage, scale: UIScreen.main.scale, orientation: .up)
        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
           let imgView: UIImageView = UIImageView.init(image: img)
           imgView.frame = CGRect.init(x: 0, y: 0, width: img.size.width, height: img.size.height)
           rootViewController.view.addSubview(imgView)
        }
    }
} catch {}
`````

Example - Error handling
````swift
do {
    _ = try person.getProperty(name: "doNotExist") //calling a property that doesn't exist
} catch let e as FREError {
    if let aneError = e.getError(#file, #line, #column) {
        return aneError //return the error as an actionscript error
    }
} catch {}
`````
----------
### Running on Simulator

iOS 11 enforces x86_64 for simulator builds. AIR only produces i386 simulator builds. Therefore, you cannot run projects with iOS11 ANEs in the XCode Simulator 

### Running on Device !!

You need to do the following first.
1) Download latest XCode 9.0 Beta from developer.apple.com
2) Install latest iOS11 Beta on your device.
3) Make a copy of your AIR SDK. We are going to be modding it.
4) You need to create a symlink from your AIRSDK to the iOS SDK   
````bash ln -s /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld /Users/User/sdks/AIR/AIRSDK_26_IOS11/lib/aot/bin/ld64/ld64````
5) Copy this tbd from here:   
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/lib/libgcc_s.1.tbd   
to here:   
/Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/lib/libgcc_s.1.tbd

6) Get the provided example project running. Don't be impatient and go gung-ho on your own project before studying the example.

The example project needs to be built and signed in the correct manner.   
AIR's ADT does not yet handle signing of the shared Swift Libraries correctly.   

An AIR based packaging tool is provided at https://github.com/tuarua/AIR-iOS-Packager     
Here is a video [demonstrating how to use it](https://www.youtube.com/watch?v=H-G8WugNFQM&feature=youtu.be)       

[![youtube video](https://raw.githubusercontent.com/tuarua/WebViewANE/master/screenshots/ios-packaging.jpg)](https://www.youtube.com/watch?v=H-G8WugNFQM&feature=youtu.be)

### Prerequisites

You will need

- Xcode 9.0 Beta
- IntelliJ IDEA
- AIR 26


CHANGELOG.md
### iOS11 Beta and XCode 9 branch
- Add PDFKit to check iOS 11 can be used
- Upgrade to Swift 4.0

### 0.0.7 
- Convert FreXxxxSwift classes to camel case.
- Change FRENamedFunction array to prevent ARC of strings.
- Prevent XCode from overwriting -Swift header. Optimised values are maintained.

### 0.0.6 
- Added getFunctions(). Functions are defined here. Removes need to edit -Swift.h bridging header
- Update to Swift 3.1 + Xcode 8.3

### 0.0.5 
- Restructure

### 0.0.4 
- Rewrite of API. Now similar to the Air Native Extension API for Android.

### 0.0.3  
- Added remaining FRE methods

### 0.0.2  
- Ported FlashRuntimeExtensions to Swift. Allows calling FRE functions from within Swift

### 0.0.1  
- initial version


native_library/ios/SwiftIOSANE/SwiftIOSANE_FW/SwiftController.swift
/*@copyright The code is licensed under the[MIT
 License](http://opensource.org/licenses/MIT):
 
 Copyright © 2017 -  Tua Rua Ltd.
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files(the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions :
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.*/

import Foundation
import CoreImage
import PDFKit

public class SwiftController: FreSwiftController {
    private var context: FreContextSwift!
    private func trace(_ value: Any...){
        freTrace(ctx: context, value: value)
    }
    
    var pdfView:PDFView!
    
    @objc public func getFunctions(prefix: String) -> Array<String> {
        
        functionsToSet["\(prefix)runStringTests"] = runStringTests
        functionsToSet["\(prefix)runNumberTests"] = runNumberTests
        functionsToSet["\(prefix)runIntTests"] = runIntTests
        functionsToSet["\(prefix)runArrayTests"] = runArrayTests
        functionsToSet["\(prefix)runObjectTests"] = runObjectTests
        functionsToSet["\(prefix)runBitmapTests"] = runBitmapTests
        functionsToSet["\(prefix)runByteArrayTests"] = runByteArrayTests
        functionsToSet["\(prefix)runErrorTests"] = runErrorTests
        functionsToSet["\(prefix)runErrorTests2"] = runErrorTests2
        functionsToSet["\(prefix)runDataTests"] = runDataTests
        
        var arr: Array<String> = []
        for key in functionsToSet.keys {
            arr.append(key)
        }
        return arr
    }

    func runStringTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start String test***********")
        guard argc == 1,
              let inFRE0 = argv[0],
              let airString: String = FreObjectSwift(freObject: inFRE0).value as? String else {
            return nil
        }

        trace("String passed from AIR:", airString)

        let swiftString: String = "I am a string from Swift"

        do {
            return try FreObjectSwift(string: swiftString).rawValue
        } catch {
        }
        return nil
    }

    func runNumberTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start Number test***********")
        guard argc == 1, let inFRE0 = argv[0],
              let airNumber: Double = FreObjectSwift(freObject: inFRE0).value as? Double else {
            return nil
        }

        trace("Number passed from AIR:", airNumber)
        let swiftDouble: Double = 34343.31

        do {
            return try FreObjectSwift(double: swiftDouble).rawValue
        } catch {
        }

        return nil
    }

    func runIntTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start Int Uint test***********")
        guard argc == 2, let inFRE0 = argv[0],
              let inFRE1 = argv[1],
              let airInt: Int = FreObjectSwift(freObject: inFRE0).value as? Int,
              let airUInt: Int = FreObjectSwift(freObject: inFRE1).value as? Int else {
            return nil
        }


        trace("Int passed from AIR:", airInt)
        trace("UInt passed from AIR:", airUInt)

        let swiftInt: Int = -666
        let swiftUInt: UInt = 888
        do {
            try _ = FreObjectSwift(uint: swiftUInt).value
            return try FreObjectSwift(int: swiftInt).rawValue
        } catch {
        }

        return nil
    }

    func runArrayTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start Array test NEW ***********")

        guard argc == 1, let inFRE0 = argv[0] else {
            return nil
        }

        let airArray: FreArraySwift = FreArraySwift.init(freObject: inFRE0)
        do {
            let airArrayLen = airArray.length

            trace("Array passed from AIR:", airArray.value)
            trace("AIR Array length:", airArrayLen)


            if let itemZero: FreObjectSwift = try airArray.getObjectAt(index: 0) {
                if let itemZeroVal: Int = itemZero.value as? Int {
                    trace("AIR Array elem at 0 type:", "value:", itemZeroVal)
                    let newVal = try FreObjectSwift.init(int: 56)
                    try airArray.setObjectAt(index: 0, object: newVal)
                    return airArray.rawValue
                }
            }

        } catch let e as FreError {
            _ = e.getError(#file, #line, #column)
        } catch {
        }

        return nil

    }

    func runObjectTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start Object test***********")

        guard argc == 1, let inFRE0 = argv[0] else {
            return nil
        }


        let person = FreObjectSwift.init(freObject: inFRE0)

        do {

            if let freAge = try person.getProperty(name: "age") {
                if let oldAge: Int = freAge.value as? Int {
                    let newAge = try FreObjectSwift.init(int: oldAge + 10)
                    try person.setProperty(name: "age", prop: newAge)

                    trace("current person age is", oldAge)

                    if let addition: FreObjectSwift = try person.callMethod(name: "add", args: 100, 31) {
                        if let sum: Int = addition.value as? Int {
                            trace("addition result:", sum)
                        }
                    }
                    if let dictionary: Dictionary<String, AnyObject> = person.value as? Dictionary<String, AnyObject> {
                        trace("AIR Object converted to Dictionary using as? Dictionary:", dictionary.description)
                    }

                    return person.rawValue

                }

            }
        } catch let e as FreError {
            _ = e.getError(#file, #line, #column)
        } catch {
        }


        return nil

    }

    func runBitmapTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start Bitmap test***********")
        guard argc == 1, let inFRE0 = argv[0] else {
            return nil
        }
        
        trace("PDFKit is iOS11 beta - iOS11 beta in AIR... Nice!!")
        pdfView = PDFView() //PDFView is iOS 11 beta !!
        pdfView.translatesAutoresizingMaskIntoConstraints = false
        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
            rootViewController.view.addSubview(pdfView)
            
            pdfView.backgroundColor = UIColor.white
            pdfView.leadingAnchor.constraint(equalTo: rootViewController.view.leadingAnchor).isActive = true
            pdfView.trailingAnchor.constraint(equalTo: rootViewController.view.trailingAnchor).isActive = true
            pdfView.topAnchor.constraint(equalTo: rootViewController.view.topAnchor).isActive = true
            pdfView.bottomAnchor.constraint(equalTo: rootViewController.view.bottomAnchor).isActive = true
            
            if let url = URL.init(string: "http://fpdownload.macromedia.com/pub/labs/flashruntimes/shared/air26_flashplayer26_releasenotes.pdf") {
                pdfView.document = PDFDocument.init(url: url)
            }
        }
        trace("pdf test finish")
        
        let asBitmapData = FreBitmapDataSwift.init(freObject: inFRE0)
        defer {
            asBitmapData.releaseData()
        }
        do {
            if let cgimg = try asBitmapData.getAsImage() {
                let context = CIContext()
                if let filter = CIFilter(name: "CISepiaTone") {
                    filter.setValue(0.8, forKey: kCIInputIntensityKey)
                    let image = CIImage.init(cgImage: cgimg)
                    filter.setValue(image, forKey: kCIInputImageKey)
                    let result = filter.outputImage!
                    if let cgImage = context.createCGImage(result, from: result.extent) {
                        let img:UIImage = UIImage.init(cgImage: cgImage, scale: UIScreen.main.scale, orientation: .up)
                        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
                            let imgView: UIImageView = UIImageView.init(image: img)
                            imgView.frame = CGRect.init(x: 10, y: 120, width: img.size.width, height: img.size.height)
                            rootViewController.view.addSubview(imgView)
                        }
                    }
                }
            }
        } catch {
        }
        

        trace("bitmap test finish")


        return nil
    }

    func runByteArrayTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start ByteArray test***********")

        guard argc == 1, let inFRE0 = argv[0] else {
            return nil
        }

        let asByteArray = FreByteArraySwift.init(freByteArray: inFRE0)

        if let byteData = asByteArray.value {
            let base64Encoded = byteData.base64EncodedString(options: .init(rawValue: 0))

            trace("Encoded to Base64:", base64Encoded)
        }
        asByteArray.releaseBytes() //don't forget to release
        return nil

    }

    func runDataTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start ActionScriptData test***********")
        if let objectAs = argv[0] {
            do {
                try context.setActionScriptData(object: objectAs)
                return try context.getActionScriptData()
            } catch {
            }
        }
        return nil
    }

    func runErrorTests(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        trace("***********Start Error Handling test***********")

        guard argc == 1,
              let inFRE0 = argv[0] else {
            return nil
        }



        let person = FreObjectSwift.init(freObject: inFRE0)

        do {
            _ = try person.callMethod(name: "add", args: 2) //not passing enough args
        } catch let e as FreError {
            trace(e.message) //just catch in Swift, do not bubble to actionscript
        } catch {
        }

        do {
            _ = try person.getProperty(name: "doNotExist") //calling a property that doesn't exist
        } catch let e as FreError {
            if let aneError = e.getError(#file, #line, #column) {
                return aneError //return the error as an actionscript error
            }
        } catch {
        }

        return nil
    }

    func runErrorTests2(ctx: FREContext, argc: FREArgc, argv: FREArgv) -> FREObject? {
        guard argc == 1,
              let inFRE0 = argv[0] else {
            return nil
        }

        let expectInt = FreObjectSwift.init(freObject: inFRE0)
        guard FreObjectTypeSwift.int == expectInt.getType() else {
            trace("Oops, we expected the FREObject to be passed as an int but it's not")
            return nil
        }

        let _: Int = expectInt.value as! Int;


        return nil

    }

    @objc public func setFREContext(ctx: FREContext) {
        context = FreContextSwift.init(freContext: ctx)
    }


}
